### ISSUE TITLE
CollisionObject2D mouse_entered  and mouse_exited events are broken

### ISSUE THREAD

**Issue by LordBrkica** [godotengine/godot#49286]
**Godot version:**
3.3.1Mono

**OS/device including version:**
Win10, rx570

**Issue description:**
CollisionObject2D has 2 events:
- mouse_entered  
- mouse_exited 

These events don't RELIABLY fire ! These events are broken ! A big warning should be standing in the documents telling this.

The events don't reliably fire in the following cases:
 - 2 or more CollisionShape2D objects are close to one another (or overlapping)
 - mouse is being moved fast between them

**Steps to reproduce:**
The events don't reliably fire in the following cases:
 - 2 or more CollisionShape2D objects are close to one another (or overlapping)
 - mouse is being moved fast between them

In my strategy game I have area2d objects and I tried using mouse_entered and mouse_exited to detect when the mouse is being hovered over them. The sad reality is this works some of the time (basically the faster you move the mouse between objects the higher the chance events wont be fired). 

Image Age of Empires where every time you move your mouse over an object there is some X chance that your hovering mouse is NOT registered. If your 2 objects (for example barracks and stable) are very close to one another the chance the events will fail is >50%.

Saddest part is I detected this bug quite late in development because with only a few objects which are far apart everything seems to be ok.

**Comment by Calinou** [comment]
> mouse is being moved fast between them

This is because CollisionObject mouse signals don't make use of continuous collision detection (CCD).

**Comment by LordBrkica** [comment]
I've just read your documentation edit and generally its ok, but here I have small remark:

> [b]Note:[/b] Due to the lack of continuous collision detection, this signal may not be emitted if the mouse moves fast enough and the [CollisionObject2D]'s **area is small.**

To be honest I don't think size has much effect (big or small) if any. A much bigger influence than size is the "empty" distance between the 2 CollisionShapes2d. The bigger the distance the more reliably do the events fire.

**Comment by pouleyKetchoupp** [comment]
@LordBrkica Could you please provide a minimal reproduction project?

I was trying to confirm it's a limitation of the physics system as opposed to a bug, and couldn't reproduce the issue so far.

Here's the project I'm testing with:
[mouse_enter.zip](https://github.com/godotengine/godot/files/6600044/mouse_enter.zip)

**Comment by LordBrkica** [comment]
I tried to start up your project but all I see is the blank screen:
![image](https://user-images.githubusercontent.com/51578785/120838580-4207a500-c568-11eb-9db7-2646effc07dd.png)

**Comment by pouleyKetchoupp** [comment]
You need to check `Debug` > `Visible Collision Shapes` in the top menu before starting to see the collision shapes in game.

**Comment by LordBrkica** [comment]
Your project works ok. I thought that C# might have an issue so I tried to make a minimal example there.
And in my minimal example here - everything works ok ! 
[MouseSignalsCSharp.zip](https://github.com/godotengine/godot/files/6600372/MouseSignalsCSharp.zip)

I spent some time thinking and now I think I understand the problem.
The "problem" is not that signals are not reliably firing, they are firing ok !

The "problem" is that I assumed one thing: **signals will always fire in the same/logical order**. And that is not the case.

If we have 2 Area2D objects close to one another but separated by some 5 pixels and NOT overlaping, it is intuitive to think that the events will be in the order:
- A mouse_entered 
- A mouse_exited
- B mouse_entered
- B mouse_exited
- A mouse_entered
-  ....

After all in nature this is the way objects interact. However in Godot it is perfectly possible to experience the following scenario:
1. enter house A
2. enter house B 
3. leave house A (this is an event that comes with lag, mouse is still at house B !!!!)

In our world its not physically possible to enter a house B 5 meters away if we first don't exit the house A.
So basically its not a bug, its just a the way how godot works.
My code assumed that the last event recieved means that the mouse has now left a house and is not hovering over any houses.

Thank you all for your help.
